### DISCUSSION TITLE
New GlobalLock SPI

### DISCUSSION THREAD

**Discussion by mhajas** [keycloak/keycloak#15138] [Archive]
### What we currently have

There is a [`DBLockProvider`](https://github.com/keycloak/keycloak/blob/19.0.3/server-spi-private/src/main/java/org/keycloak/models/dblock/DBLockProvider.java) interface that allows us to acquire [the following locks](https://github.com/keycloak/keycloak/blob/19.0.3/server-spi-private/src/main/java/org/keycloak/models/dblock/DBLockProvider.java#L45-L47).

### New storage

The new storage split Keycloak data into independent areas, where each area can be initialized independently. For example, it is not needed to lock the whole database when initializing the authorization services area.

The current interface is not flexible enough to provide a way to simply introduce a new block of code that is synchronized across the whole cluster. 

### Suggested changes

We suggest adding a new `GlobalLockSPI` that allows us to request a new `GlobalLock` for any `String` identifier that should be replicated across the whole cluster.

[Here is a PR](https://github.com/keycloak/keycloak/pull/14751) with the SPI introduction and `DBLock` compatible implementation for that SPI. After the PR is merged we will introduce implementation for [HotRod](https://github.com/keycloak/keycloak/issues/14721) and [JPA](https://github.com/keycloak/keycloak/issues/14734) map storages. 

Any thoughts or ideas for improvement?

**Comment by mhajas** [comment]
The current state of the GlobalLock API suggests using `try-with-resources` for releasing locks.

Based on the use case it may not be enough to synchronize just one block of code inside `try-with-resources`, for example, in case we want to update some storage as part of the locked period of time (in that case we need to release the lock after the data are changed in the store). 

This raises a new set of questions to answer before we are able to continue providing lock implementations. 
1. Do we need to consider storage updates as part of the locked period of time or do we want to use `GlobalLock` API only in case it is possible to lock one block of code?
2. If we don't want to consider storage changes for this API, will we have another way for doing this? Theoretically, we could consider each storage to use some internal implementation of locking to achieve this. For example, HotRod storage can use `locks` cache directly without using GlobalLockAPI for pessimistic locking, or JPA can use `SELECT FOR UPDATE` for schema update.
3. If we want to consider storage changes, how can we make sure the locks are released at the correct time? How can we avoid a situation where, for example, JPA `commit()` releases a lock, but some HotRod storage changes that should be synchronized by the lock are not yet committed to the Infinispan? This issue can be even more visible at the moment because we don't have XA transactions for all stores and we also have this issue #13575.

**Reply by ahus1** [reply]
Thanks for the summary, I think it states the problems and shows some solutions. 

My view on "commit change to store before or together with the commit": it is the general approach, for example when acquiring the lock to run a cleanup Job, or to setup the master realm.

So I suggest:

- Never use try-with-resources, instead always free the locks together with the commit, as this covers both the data changing scenarios and the non-data-changing scenarios. If we want a different, more performant locking type, we could add it once the performance tests show the need.
- Change all (?) areas of `enlist()` and use XA for all transactions which can participate (LDAP will never, as we all know), and then assume that locks are released as part of the XA transaction across the different stores. Maybe we could eventually squeeze LDAP transactions in between the prepare and commit phase of XA transactions. Not using XA feels like reinventing the wheel. If performance tests show that this is too expensive there might be an option later to disable it, but then that person will need to live the the trade-off of inconsistencies.

**Reply by hmlnarik** [reply]
Thank you for the suggestion.

This argumentation builds on the assumption that the commit would always release the lock. However, `GlobalLock` API is a global lock which lasts across transactions, hence the `commit` cannot release the lock. For example database creation needs a lock that survives several commits: The tables are created and data filled in more than one transaction; DDL statements often implicitly commit. Please see below for further arguments.

**Comment by hmlnarik** [comment]
> 1. Do we need to consider storage updates as part of the locked period of time or do we want to use GlobalLock API only in case it is possible to lock one block of code?

Yes, we need it. This is the purpose for  `GlobalLock` replacing the `DBLockProvider` API.

Item 2 condition is invalidated by the statement above.

> 3. If we want to consider storage changes, how can we make sure the locks are released at the correct time? How can we avoid a situation where, for example, JPA commit() releases a lock, but some HotRod storage changes that should be synchronized by the lock are not yet committed to the Infinispan? This issue can be even more visible at the moment because we don't have XA transactions for all stores and we also have this issue Replace enlistAfterCompletion with enlist for ex-legacy-infinispan areas #13575.

The argument that _JPA `commit()` releases a lock [before an action is finished]_ is not correct. If such a global lock is needed, then the lock itself has to be made visible, i.e. has to run in its own transaction and has to be persisted. The pseudocode is as follows:

1. In a separate transaction, obtain the lock and a `GlobalLock` instance (`unlock`) to unlock it.
2. In a separate transaction, perform all changes that require this lock, e.g. create database, initialize a realm.
3. In a separate transaction, execute `unlock` code from step 1.

This can be rewritten as 
```java
  try (GlobalLock unlock = lock.acquire("lockName")) {
    KeycloakModelUtils.runJobInTransaction(job);
  }
```

Note that both `lock.acquire()` and the `close` method could leverage `KeycloakModelUtils.runJobInTransaction` which would ensure separation of transactions without boilerplate code.

The usecase for short-lived `try-with-resources` locks is complementary to the long-lived one from above. It is used for cases like locking a single object against concurrent changes (e.g. via `SELECT FOR UPDATE`). This is usually an implementation detail of a particular implementation, and is used for objects within it. There is not much use for explicit use of the `GlobalLock` API there, it is possible though that the same mechanism as driving the global locks would (**not *must***) be used there.

This can be illustrated on a JPA example. (Please do not treat this as an ultimate solution.) `SELECT FOR UPDATE` lock is released with commit. Thus it is insufficient for long-term locks. However, the acquire-lock phase in JPA can perform both `SELECT FOR UPDATE` to some _globalLockAcquirer_ lock and insertion of a lock identifier into a lock table; the unlocking mechanism would remove the inserted row.

**Reply by ahus1** [reply]
Thank you for this response, I can now see that there has been a misunderstanding in a previous discussion when there was the question if a lock can live longer than the current transaction. The answer was then "no", and so we it was modelled as part of the current transaction to clean things up. 

From your reply I understand that the locks live outside of the current transaction (as this is the only way to unlock them later without committing the current transaction). And I also see that running any DB modifications can only run in another nested transaction as there is no other to commit the changed data before releasing the lock - like you showed in the snippet. Happy to see that as part of the JavaDoc eventually. 

Again, thanks for the clarification.

**Comment by mhajas** [comment]
We discussed this offline with @hmlnarik and we came up with the following:

The purpose of GlobalLock API is to provide a way to synchronize some code execution across more Keycloak nodes/threads. The locks can be used in 3 situations:

### 1
An execution with all its effects being visible to others can be enclosed into one block of code. In this situation, we don't need to take transactional processing into account as all effects of the execution are visible to everyone at the time the code block is finished. For this purpose, we recommend using the `try-with-resources` block. For example:
```java
try (GlobalLock lock = lockProvider.acquire("DATABASE_UPDATE_LOCK")) {
    // update database schema if necessary
}
```
When this execution is finished the schema is updated and all other nodes can see the new schema. The lock is released.

### 2
Effects of an execution that is synchronized are visible only when the current transaction commits. In this situation, we can not use `try-with-resources` block because, at the end of the try block, the changes are not visible for other nodes/threads yet, therefore we need to unlock the lock after the changes are committed. As we can not force some specific order of `commit` executions of enlisted Keycloak transactions, we can't use `KeycloakTransactionManager` for releasing the lock, we need to make sure all transactions are committed or rolled-back before we release the lock. For this purpose, we ask implementors of the API to keep track of all provided locks and release them at `KeycloakSession.close` phase. 

An example of this situation is when we need to synchronize some execution that triggers a change in Keycloak storage.
```java
lockProvider.acquire("MY_LOCK_NAME")
RealmModel realm = session.getRealm("realmName");
if (!realm.getAttributes().contains("myAttrThatIsCreatedOnlyByOneInstance")) {
   realm.addAttribute("myAttrThatIsCreatedOnlyByOneInstance", "I created it " + keycloakInstanceIdentifier());
}
```
At the end of this block of code, the lock cannot be released because the changes done by `addAttribute` are not visible to other nodes yet. The lock will be released when the current `session` ends.

### 3
The execution is done in more than one transaction and it needs to be acquired until all of them are finished. An example of this situation is when we want to run some migration from one store to another in more bulks. Each bulk is migrated in one of the transactions and the lock can be released only when all bulks were migrated.

For this purpose we can use approach in point 1, as we can do something like this:
```java
try (GlobalLock lock = locksProvider.acquire("MIGRATON_LOCK")) {
    for (Bulk b : bulks) {
        runJobInTransaction() {
            b.execute();
        }
    }
}
```

Here is a Javadoc of GlobalLockProvider suggestion based on the above: https://gist.github.com/mhajas/ae24e0c19abe20d6430457a24fce26a3
